index.bs

```diff
@@ -976,319 +976,352 @@ their contract with the developer.
+For example, DOM {{Node}}s are live objects,
+to allow the author to make changes to the document
+with an understanding of the current state.
+
+Properties of live objects may be computed as they are accessed,
+instead of when the object is created.
+This makes live objects sometimes a better choice
+if the data needed is complex to compute,
+since there is no need to compute all the data before the object is returned.
+
+A live object may also use less memory,
+since there is no need to copy data to a static version.
+
+<h4 id="static-objects" class="no-num no-toc">Static objects</h4>
+
+If an object represents a list that might change,
+most often the object should be static.
+This is so that code iterating over the list
+doesn't need to handle the possibility of the list changing in the middle.
+
+<div class="example">
+{{Document/getElementsByTagName()}} returns a live object which represents a list,
+meaning that authors need to take care when iterating over its items.
+
+<!-- does anyone know more specifics here? -->
+The choice to have {{ParentNode/querySelectorAll()}} return static objects
+was made after spec authors noticed that {{getElementsByTagName()}}
+was causing problems.
+
+{{URLSearchParams}} isn't static,
+even though it represents a list,
+because it's the way for authors to change the query string of a URL.
+</div>
+
+Note:  For <a>maplike</a> and <a>setlike</a> types,
+this advice may not apply,
+since these types were designed to behave well
+when they change while being iterated.
+
+If it would be complex to keep the data needed by an object up to date,
+a static object avoids having to keep the object updated
+until it's garbage collected,
+even if it isn't being used.
+<!-- why wouldn't it be computed lazily though? -->
+
+If a static object represents some state which may change frequently,
+it should be returned from a method,
+rather than available as an attribute.
+
+See also:
+
+* [[#attributes-like-data]]
+
+<h3 id="optional-parameters">Make function parameters optional if possible</h3>
+
+If a parameter for an API function has a reasonable default value,
+make that parameter optional and specify the default value.
+
+<div class="example">
+{{EventTarget/addEventListener()}} takes an optional boolean `useCapture` argument.
+Thie defaults to `false`, meaning that
+the event should be dispatched to the listener in the bubbling phase by default.
+</div>
+
+Note: Exceptions have been made for legacy interoperability reasons
+(such as {{XMLHttpRequest}}),
+but this should be considered a design mistake rather than recommended practice.
+
+The API must be designed so that if a parameter is left out,
+the default value used is the same as
+converting `undefined` to the type of the parameter.
+For example, if a boolean parameter isn't set,
+it must default to false.
+
+See also:
+
+* [[#prefer-dict-to-bool]]
+
+<h3 id="prefer-dict-to-bool">Prefer dictionary parameters over primitive parameters</h3>
```

Comment:
Maybe this principle should come before `#optional-parameters`?

Reply:
Done.